pulsar-broker/src/main/java/org/apache/pulsar/broker/loadbalance/extensions/BrokerRegistryImpl.java

```diff
@@ -0,0 +1,218 @@
+                    return null;
+                });
+        return future;
+    }
+
+    @Override
+    public CompletableFuture<Optional<BrokerLookupData>> lookupAsync(String broker) {
+        return brokerLookupDataLockManager.readLock(keyPath(broker));
+    }
+
+    @Override
+    public void forEach(BiConsumer<String, BrokerLookupData> action) {
+        this.brokerLookupDataMap.forEach(action);
+    }
+
+    public void listen(BiConsumer<String, NotificationType> listener) {
+        this.listeners.add(listener);
+    }
+
+    @Override
+    public void close() throws PulsarServerException {
+        try {
+            this.unregister();
+        } catch (MetadataStoreException ex) {
+            if (ex.getCause() instanceof MetadataStoreException.NotFoundException) {
+                throw new PulsarServerException.NotFoundException(MetadataStoreException.unwrap(ex));
+            } else {
+                throw new PulsarServerException(MetadataStoreException.unwrap(ex));
+            }
+        }
+
+        this.listeners.clear();
+    }
+
+    private void handleMetadataStoreNotification(Notification t) {
+        if (t.getPath().startsWith(LOOKUP_DATA_PATH) && t.getPath().length() > LOOKUP_DATA_PATH.length()) {
+            try {
+                if (log.isDebugEnabled()) {
+                    log.debug("Handle notification: [{}]", t);
+                }
+                this.scheduler.submit(() -> {
+                    String lookupServiceAddress = t.getPath().substring(LOOKUP_DATA_PATH.length() + 1);
+                    this.updateBrokerLookupDataToLocalCache(lookupServiceAddress, t.getType());
+                    for (BiConsumer<String, NotificationType> listener : listeners) {
+                        listener.accept(lookupServiceAddress, t.getType());
+                    }
+                });
+            } catch (RejectedExecutionException e) {
+                // Executor is shutting down
+            }
+        }
+    }
+
+    private void updateBrokerLookupDataToLocalCache(String lookupServiceAddress, NotificationType type) {
+        switch (type) {
+            case Created, Modified, ChildrenChanged -> {
+                try {
+                    Optional<BrokerLookupData> lookupData =
+                            brokerLookupDataLockManager.readLock(keyPath(lookupServiceAddress))
+                                    .get(conf.getMetadataStoreOperationTimeoutSeconds(), TimeUnit.SECONDS);
+                    if (lookupData.isEmpty()) {
+                        brokerLookupDataMap.remove(lookupServiceAddress);
+                        log.info("[{}] Broker lookup data is not present", lookupServiceAddress);
+                        break;
+                    }
+                    brokerLookupDataMap.put(lookupServiceAddress, lookupData.get());
+                } catch (Exception e) {
+                    log.warn("Error reading broker data from cache for broker - [{}], [{}]",
```

Comment:
In this case, this `brokerLookupDataMap` cache can be inconsistent from others. Can we make this cache update eventually consistent?